File: luci-app-photondns-compat/root/usr/lib/lua/luci/model/cbi/photondns/rules.lua
-- Rules (CBI) — legacy port of rules.js
-- Four file-backed text areas; photondns is restarted on save.
local fs = require "nixio.fs"

local RULE_FILES = {
	{ "hosts", "/etc/photondns/hosts.txt", translate("Hosts"),
		translate('Static host records: "<name> <ip> [ip...]" per line. Answers A/AAAA locally.') },
	{ "block", "/etc/photondns/block.txt", translate("Block List"),
		translate('Domains answered with NXDOMAIN. "example.com" blocks all subdomains, "full:example.com" exact only.') },
	{ "local_domains", "/etc/photondns/local_domains.txt", translate("Local Domains"),
		translate('Domains resolved by the "local" upstream group (configure Local-domain DNS servers first).') },
	{ "redirect", "/etc/photondns/redirect.txt", translate("Redirect"),
		translate('"<from-domain> <to-domain>" per line: answer queries for from-domain with the records of to-domain.') },
	{ "prewarm", "/etc/photondns/prewarm.txt", translate("Prewarm"),
		translate('Domains kept always-resolved (one per line) so a first visit is never a slow cold miss. Default set = YouTube/Google. Enable "Prewarm popular domains" in Basic Settings.') },
	{ "lan_hosts", "/etc/photondns/lan_hosts.txt", translate("LAN Hosts"),
		translate('Extra LAN hosts pinned by name → IP, one "name ip [ip...]" per line, for devices that do not advertise a usable DHCP name (e.g. a Mac). Each resolves bare and under the LAN suffix, and answers reverse PTR. DHCP leases are learned automatically; this file is only for pins. Enable "Resolve LAN hostnames" in Basic Settings.') },
}

local m, s, o

m = Map("photondns", translate("photondns Rules"),
	translate("Rule files are applied on service restart."))
m.apply_on_parse = true

s = m:section(NamedSection, "main", "photondns")
s.addremove = false
s.anonymous = true

for _, f in ipairs(RULE_FILES) do
	local key, path, title, descr = f[1], f[2], f[3], f[4]
	s:tab(key, title)
	o = s:taboption(key, TextValue, "_" .. key, nil, descr)
	o.rows = 20
	o.wrap = "off"
	o.rmempty = true
	o.cfgvalue = function(self, section)
		return fs.readfile(path) or ""
	end
	o.write = function(self, section, value)
		value = (value or ""):gsub("\r\n", "\n")
		value = value:gsub("^%s+", ""):gsub("%s+$", "") .. "\n"
		fs.writefile(path, value)
	end
	o.remove = function(self, section)
		fs.writefile(path, "\n")
	end
end

function m.on_after_commit(self)
	luci.sys.call("/etc/init.d/photondns restart >/dev/null 2>&1")
end

return m

File: luci-app-photondns/htdocs/luci-static/resources/view/photondns/rules.js
'use strict';
'require form';
'require fs';
'require ui';
'require view';

const RULE_FILES = [
	['hosts', '/etc/photondns/hosts.txt', _('Hosts'),
		_('Static host records: "<name> <ip> [ip...]" per line. Answers A/AAAA locally.')],
	['block', '/etc/photondns/block.txt', _('Block List'),
		_('Domains answered with NXDOMAIN. "example.com" blocks all subdomains, "full:example.com" exact only.')],
	['local_domains', '/etc/photondns/local_domains.txt', _('Local Domains'),
		_('Domains resolved by the "local" upstream group (configure Local-domain DNS servers first).')],
	['redirect', '/etc/photondns/redirect.txt', _('Redirect'),
		_('"<from-domain> <to-domain>" per line: answer queries for from-domain with the records of to-domain.')],
	['prewarm', '/etc/photondns/prewarm.txt', _('Prewarm'),
		_('Domains kept always-resolved (one per line) so a first visit is never a slow cold miss. Default set = YouTube/Google. Enable "Prewarm popular domains" in Basic Settings.')]
];

return view.extend({
	load() {
		return Promise.all(RULE_FILES.map(f =>
			L.resolveDefault(fs.read(f[1]), '')
		));
	},

	render(data) {
		let m, s, o;

		m = new form.Map('photondns', _('photondns Rules'),
			_('Rule files are applied on service restart.'));

		s = m.section(form.NamedSection, 'main', 'photondns');

		RULE_FILES.forEach((f, i) => {
			s.tab(f[0], f[2]);
			o = s.taboption(f[0], form.TextValue, '_' + f[0], null, f[3]);
			o.rows = 20;
			o.cfgvalue = () => data[i] || '';
			o.write = (section_id, value) => {
				const cur = data[i] || '';
				const next = (value || '').trim().replace(/\r\n/g, '\n') + '\n';
				if (cur === next) return;
				return fs.write(f[1], next);
			};
			o.remove = () => fs.write(f[1], '\n');
		});

		return m.render();
	},

	handleSaveApply(ev) {
		return this.handleSave(ev).then(() => {
			return fs.exec('/etc/init.d/photondns', ['restart']);
		}).then(() => {
			ui.addNotification(null, E('p', _('Rules saved and photondns restarted.')), 'info');
		});
	}
});
